Add HashValidator to validate types in config arrays

othercorey opened the pull request:
Returns an array of errors for each field in the array with the dot notation path to the field as the key. All fields in an array are validated even if one has an error.

## Schema Specification

Example: 

```php
$schema = [
    'strict' => false, // optional, defaults to true
    'required' => true, // optional, defaults to false
    'allowPaths' => false, // optional, defaults to true
    'fields' => [
        'field' => field-specification
    ],
]
```

The `strict` and `required` flags are optional while the fields array must contain at least 1 field.

- `strict`

    The `strict` flag controls whether all of the keys in the data must exist in the schema. Disabling this essentially allows for 
    undocumented pass-through fields.

    This can be overridden for any `array{}` type field 

    Defaults to `true`.

- `required`

    The `required` flag controls whether the field must exist in the data. Enabling essentially means there are no optional fields in the schema.

    This can be overridden for any field.

    Defaults to `false`.

- `allowPaths`

  The `allowPaths` flag controls whether dot notation paths are expanded in in the top-level fields of the hash. Paths are not parsed in nested fields.

  Defaults to true.

## Field Specification

Example: 

```php
$schema = [
    'fields' => [
        'field' => 'int',
        'union_field' => [
           'required' => true,
           'type' => ['int', 'bool'],
        'array_field' => [
            'strict' => false,
            'fields' => [
              ...
           ],
       ],
    ],
]
```

The field specification can be either a string on an array. If a string, that is used as the type for the field.  If an array, then the `type` key is required can be either a string or array.

The `required` flag can be set for each field.

The `type` key can be either a single basic type or an array of basic, generic and array shapes. Generic types and array shapes are arrays with the type name as the key.

```php
[
    type => ['string', 'class-string<>' => stdClass::class, 'array{}' => ['fields' => [...]]
]
```

When the type is an array, it acts as a union type where the field must be one of the types.

Basic Types:

- 'array'
- 'list'
- 'string'
- 'int'
- 'float'
- 'bool'
- 'null'
- 'mixed'

Generic Types:

- ['array<>' => string]
- ['list<>' => string]
- ['class-string<>' => string]

Array Shape:
- ['array{}' => ['fields' => fields]

### Array Shape Specification

Array shapes are represented by the `array{}` type and must be an array that contains a `fields` array the same as the top-level schema.

```php
[
    'type' => [
        'array{}' => [
            'strict' => false,
            'fields' => [
               'array_field' => 'int',
            ],  
        ], 
    ],
]
```

The `strict` flag overrides the schema default for that array only.

othercorey committed:
Add HashValidator to validate types in config arrays

markstory commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
+
+            foreach ($keyTypes as $keyType) {
+                if (static::validateValue($keyType, $value)) {
+                    continue 2;
```
I'm wary of using `continue 2`. Should we have a protected method instead for this inner loop?

othercorey replied:
If we supported typed/nested/shaped arrays, this would likely have to change anyway.

markstory commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
+     * @param bool $strict Require all keys in both $data and $types to match
+     * @return array
+     */
+    public static function validate(array $data, array $types, bool $strict = false): array
```
I like the general idea of being able validate the 'shape' of an array. Should we consider using a format like jsonschema or is that too heavy handed? Right now a limitation I can see is that this can't be used for nested arrays which we have a bit of in our configuration data.

othercorey replied:
Nested arrays would require nested types. At the moment, I am using arrays to store more than one allowed value: `['string', 'bool'].  That would need to change into something more structured. It's something I want to support, but stopped before getting feedback.

ADmad commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
+     */
+    protected static function validateValue(string $type, $value): bool
+    {
+        if ($type === 'null') {
```
A switch case would be nicer to read.

othercorey replied:
I had limited choice with 7.4.

othercorey replied:
Wait, sorry, I was thinking match since I originally wrote that. Yea, we can use a switch sine the type is always a string.

othercorey commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
+            return is_array($value);
+        }
+        if ($type === 'string') {
+            return is_string($value);
```
We could support class-string using `is_a()`.

othercorey commented:
Updated description with full specification info.

Added support for shaped arrays.

~Added some schema/specification validation.~

Reduced validation to simplify code. Can provide a separate function to check schema if needed. Only the top-level schema is validated now. Everything else should throw a PHP error.

othercorey commented:
~Would you prefer a single error string per key or is the separate `expected` and `actual` values useful? I suspect users will only need a description instead of access to the raw types (including null when missing).~

Reduced all errors to string descriptions.

othercorey commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,211 @@ public static function mergeDiff(array $data, array $compare): array
+                    }
+                    break;
+                case 'list<>':
+                    break;
```
`list<>` not yet implemented. Will only support basic types like `string` and `bool`, but not `array<>`.

othercorey commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,226 @@ public static function mergeDiff(array $data, array $compare): array
+        }
+
+        foreach ($fieldSpec['type'] as $typeKind => $type) {
+            if (is_string($typeKind)) {
```
This is required because in PHP 7.4, 0 == 'array<>' is true so can't use switch() with integer and string keys.
https://3v4l.org/TAHQh

ADmad commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,224 @@ public static function mergeDiff(array $data, array $compare): array
+     */
+    public static function validate(array $data, array $schema, bool $strict = true): array
+    {
+        $fields = $schema['fields'] ?? [];
```
Do you forsee requiring other keys besides `fields` in future?

othercorey replied:
Only reason `fields` exists now is to allow for a `version`. If we don't have an outer array then I can't tell the difference between an array of fields with a `version` field or the schema itself.

ADmad replied:
Sounds good, better to keep a provision for more keys. I wonder then if we could allow a global `required` key too instead of having to specify it for each field. That could also allow doing away with the `$strict` argument.

othercorey replied:
We could. 

`$strict` basically allows pass-through data. `required` flag on the schema itself might feel like we're saying the default `required` setting for all the fields is that value. 

Do you think having a `strict` key in the schema would make more sense?

ADmad replied:
> required flag on the schema itself might feel like we're saying the default required setting for all the fields is that value.

That's what I was suggesting. Guess I misunderstood what `$strict` did.

> Do you think having a strict key in the schema would make more sense?

Yes, I would prefer if the schema was self contained and we didn't have to rely on a special external flag to control how the validation works.

othercorey replied:
I'll add it. The original implementation tried to be super minimal by guessing what the arrays were so top-level settings didn't exist. However, that was a nightmare to write.

othercorey replied:
`strict` and `required` are now supported at the schema level. The `strict` flag can be overwritten per array shape `array{}` and `required` can be overwritten per field.

ADmad commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,224 @@ public static function mergeDiff(array $data, array $compare): array
+    protected static function validateArray(array $data, array $arraySpec, string $fieldPrefix, bool $strict): array
+    {
+        $fields = $arraySpec['fields'] ?? [];
+        if (!is_array($fields) || empty($fields)) {
```
The existence of non empty `fields` array is already checked in the calling public method, so repeating again in this protected method seems redundant.

ADmad replied:
I see you already removed it.

markstory commented on tests/TestCase/Utility/HashTest.php:
```diff
@@ -759,6 +761,344 @@ public function testMergeVariadic(): void
+    public function testValidateArrayTypes(): void
+    {
+        // test single-dimension
+        $errors = Hash::validate(
+            ['has_array' => ['int' => 1]],
+            [
+                'fields' => [
+                    'has_array' => [
+                        'type' => [
+                            'array{}' => [
+                                'fields' => ['int' => 'int'],
+                            ],
+                        ],
+                    ],
+                ],
+            ]
+        );
+        $this->assertEmpty($errors);
```
Are there any lessons we can take from openschema's definition format? Having an interface that reuses key terms from a more broadly used specification could help build familiarity with what we're providing.

othercorey replied:
I'm not familiar with openschema. Can you link it?

markstory replied:
Derp, I meant to say [open api schema](https://swagger.io/specification/)

othercorey replied:
The key thing we're missing compared to open api and json schema is extra type information.

json schema has a union type but open api doesn't. However, they both have these schema compositions with `anyOf` that takes an array of type definitions.

I'd like to stay close to accepted php docblock and type formats where possible.

markstory commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,179 @@ public static function mergeDiff(array $data, array $compare): array
+     * @param array $schema Schema to validate against
+     * @return array
+     */
+    public static function validate(array $data, array $schema): array
```
I like this feature quite a bit, and think it has potential to help us build safer applications. I wonder if this shouldn't be its own class though? I know `Hash` has many oddly shaped responsibilities but I think this is substantial enough that we could give it some more space.

othercorey replied:
It kind of makes sense as a Hash function because we don't support lists as the top-level array. However, there are many protected methods used to implement it.

I thought about making it a separate repo even, but cakephp code will be tied to a specific schema version. 

What would we call it?

markstory replied:
> However, there are many protected methods used to implement it.

Yeah the number of protected methods make me think that it could be a separate class.

> What would we call it?

Some possible options I can think of:

* ShapeValidator
* SchemaValidator
* ShapeValidation

We could continue to have `Hash::validate()` or `Validation::arrayShape()` and then delegate to the separate class.

markstory commented on src/Utility/Hash.php:
```diff
@@ -1146,6 +1147,179 @@ public static function mergeDiff(array $data, array $compare): array
+                /** @var class-string $elements */
+                return is_a($value, $elements, true);
+            case 'list<>':
+                break;
```
Could this be `return is_array($value) && $value == array_values($value)`. Numerically indexed arrays are about as close as we can get to 'lists'.

othercorey replied:
I'm going to check the type of each entry for the generic one `list<>`.  I'll add a basic `list` type as well that just verifies indexed arrays.

othercorey commented:
Implemented `list<>`. Added `list` and `array<>`. `array<>` is a typed array that supports any keys without requiring a specific array shape.

othercorey commented:
@cakephp/core Are there any edge cases for your config arrays that should be covered?

othercorey commented:
Moved code into separate class: `HashValidator`.

Added support for dot notation paths in the top-level hash fields.

Improved support for `list<>` and `array<>` that now accept generic types such as `class-string<>`.

othercorey commented:
Added support for array shape references. References are similar to yaml references where an existing array shape specification can be re-used.

The shared specifications must be part of the top-level `refs` array. Instead of specifying `fields` for an array shape, a `ref` key should be set to the name of the shared specification.

```php
[
    'refs' => [
        'int_shape_ref' => [
            'fields' => ['nested' => 'int'],
        ],
    ],
    'fields' => [
        'is_array' => [
            'type' => [
                'array{}' => [
                    'ref' => 'int_shape_ref',
                ],
            ],
        ],
    ],
];
```

We could make this a little conventions based where the array{} key is assigned a string that should be a reference. We could also enforce the reference name have a `&` prefix.

```php
[
    'refs' => [
        'int_shape_ref' => [
            'fields' => ['nested' => 'int'],
        ],
    ],
    'fields' => [
        'is_array' => [
            'type' => [
                'array{}' => '&int_shape_ref',
            ],
        ],
    ],
];
```

othercorey commented:
Will open a new PR with final spec and implementation.
